### ISSUE TITLE
Doctests don't work in bin targets, non-public items

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust#50784]
It's surprising that doctests sometimes don't run at all, and there's no warning about this.

```sh
cargo new foo --bin
```

```rust
/// ```rust
/// assert!(false);
/// ```
fn main() {
    println!("Hello, world!");
}
```

```sh
cargo test --all
```

I'd expect the above to fail, but it looks like the doc-comment code is never ran. Even if I explicitly enable doctests for the binary:

```toml
[[bin]]
name = "foo"
path = "src/main.rs"
doctest = true
```

they still aren't run.

(moved from https://github.com/rust-lang/cargo/issues/5477)

**Comment by GuillaumeGomez** [comment]
It sounds weird to have documentation examples in a bin target, don't you think? There is no documentation to generate so why would there be documentation's code to run?

**Comment by kornelski** [comment]
Cargo makes new users create bin targets by default, so the first time a new Rust user tries doc tests, it's probably in a bin.

This bug is based on a [feedback from a user](https://users.rust-lang.org/t/ux-feedback-from-a-rust-newbie/17220/6).

**Comment by GuillaumeGomez** [comment]
Maybe we give bad information ahead? I wonder if users know the difference between comments and doc comments... For now I don't see this as a rustdoc bug but as a misunderstanding that should be resolved in tutorials/books.

cc @rust-lang/docs

**Comment by kornelski** [comment]
Wow, that's a nice surprise that it works. I was afraid that duplicate main would be a problem, but I guess rlib namespaces it!

**Comment by QuietMisdreavus** [comment]
Yeah, the main in your bin goes inside the lib crate, so it's just a regular function at that point. I did have to tag it with `#[allow(dead_code)]` to silence a warning, but otherwise it's all good.

**Comment by PhilipDaniels** [comment]
I just hit this today. I came here after finding https://github.com/rust-lang/cargo/issues/2009 in a Google search, which mentions that it is 'addressed in the documentation'. I suspect it may have been at one point but it doesn't appear to be now. I checked

https://doc.rust-lang.org/book/ch11-01-writing-tests.html

and from there the link

https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#documentation-comments-as-tests

(Of course it's possible its mentioned somewhere and I missed it. A very careful reading of the surrounding text *implies* it, because it's all talking about libs, but nowhere I see is it mentioned explicitly.)

**Comment by joshtriplett** [comment]
I ran into this today, and found it confusing. I would love to be able to write doctests in the modules of my binary crate.

What would it take to make that work out-of-the-box?

**Comment by ehuss** [comment]
I can think of a few hurdles that someone would need to address:

* How to resolve symbols in a binary?  Doctests behave as-if they are a separate crate linking in the library.  But a binary is normally not used a library, so it has no name to refer to.  For example:

```rust
/// ```
/// do_something();
/// ```
fn do_something() {}

fn main() {}
```

Here, `do_something();` would fail because it is not in scope.  You could try to scope it with the name of the binary (`mything::do_something();`), but then you have two new problems: what if you have a lib target of the same name?  And, deal with the fact that you normally don't export any public items in a binary.

* How to deal with lack of public exports?  Building a binary as a library will result in a lot of unused warnings.  And since rustdoc imports the crate, usually nothing will be publicly exported, so you won't be able to access anything.

One (rough) idea to address this would be to fundamentally change how `rustdoc --test` works. Instead of creating a synthetic test that links to a library, embed the tests directly in the module.  So it would rewrite the above example to:

```rust
fn do_something() {}

#[test]
fn do_something_doctest() {
    do_something();
}
```

So all local (private) symbols would be in scope.

That's a pretty radical difference, so I'm not sure if it would fly (would probably be confusing if there are two styles of doctests).

Also, I think it's worth examining if this is a good idea from the start.  Documenting binaries still seems like a strange concept to me.  Why can't you just move all the code to a library?

**Comment by joshtriplett** [comment]
On Sat, Mar 28, 2020 at 02:23:19PM -0700, Eric Huss wrote:
> I can think of a few hurdles that someone would need to address:
> 
> * How to resolve symbols in a binary?  Doctests behave as-if they are a separate crate linking in the library.  But a binary is normally not used a library, so it has no name to refer to.  For example:
> 
> ```rust
> /// ```
> /// do_something();
> /// ```
> fn do_something() {}
> 
> fn main() {}
> ```
> 
> Here, `do_something();` would fail because it is not in scope.  You could try to scope it with the name of the binary (`mything::do_something();`), but then you have two new problems: what if you have a lib target of the same name?  And, deal with the fact that you normally don't export any public items in a binary.

> * How to deal with lack of public exports?  Building a binary as a library will result in a lot of unused warnings.  And since rustdoc imports the crate, usually nothing will be publicly exported, so you won't be able to access anything.
> 
> One (rough) idea to address this would be to fundamentally change how `rustdoc --test` works. Instead of creating a synthetic test that links to a library, embed the tests directly in the module.  So it would rewrite the above example to:
> 
> ```rust
> fn do_something() {}
> 
> #[test]
> fn do_something_doctest() {
>     do_something();
> }
> ```
> 
> So all local (private) symbols would be in scope.

I would love to see that as the *normal* approach to doctests,
personally. It would also unify the different kinds of tests. I do agree
that that would be a more invasive change, though.

> Also, I think it's worth examining if this is a good idea from the start.  Documenting binaries still seems like a strange concept to me.  Why can't you just move all the code to a library?

Why should the concept of documentation and testing be limited to
libraries, and not supported in binary crates? I use documentation in
binary crates to document the code for later understanding. I'd like to
use tests to verify individual modules of functionality; that doesn't
necessarily mean I want to split that functionality into a separate
library crate.

**Comment by dkasak** [comment]
Just hit this today. I also frequently document internal modules of binary crates and intuitively expected that doctests would still work. Effectively, I use internal modules as non-public libraries. The reason I do not make them into separate library crates is that they are often not useful in a general context, but I do sometimes end up turning them into libraries. This kind of flexibility seems like a win.

**Comment by ayroblu** [comment]
Hey, so I just learnt about this while learning about the language. It's disappointing, and I'm not sure this has been prioritised so it doesn't sound like it's going to be addressed in the near future.

Anyways, I guess the only way then is to do the plebeian way and write a test for it

**Comment by GuillaumeGomez** [comment]
You can split your crate into two parts: the binary part (containing only the `main` function and the args parsing eventually) which calls all other functions and the "lib" part. Then you can document all functions inside the lib part. I did that for a few crates of mine and I think it's more than enough and is a better match for how doc should be handled. What do you think?

**Comment by ayroblu** [comment]
@GuillaumeGomez Do you have a good example or minimum POC with like `cargo build`? Sorry still learning here

**Comment by GuillaumeGomez** [comment]
Sure, here you go: https://github.com/gtk-rs/gir/blob/master/Cargo.toml

**Comment by slightlytwisted** [comment]
> It sounds weird to have documentation examples in a bin target, don't you think? There is no documentation to generate so why would there be documentation's code to run?

I think mostly for consistency. If someone is used to writing docs with rustdoc and doctests on their library functions, it makes sense that they would want apply the same documentation style to their bin functions. On a small project, having to separate out helper functions to a separate library is an extra hoop to jump through, especially if it is just for the sake of getting tests to run.

I think underneath this is the assumption that people don't need to generate documentation for binaries. But in a team environment, its a good idea to document as much as you can, including examples. Having to switch documentation and test styles between libraries and binaries seems like an arbitrary constraint.

**Comment by dkasak** [comment]
> But in a team environment, its a good idea to document as much as you can, including examples.

Yes, I even document extensively for myself! And you can never be sure whether someone will join you on a project in the future.

**Comment by steveklabnik** [comment]
> I think underneath this is the assumption that people don't need to generate documentation for binaries.

It's not quite this, exactly. Rustdoc is a tool for documenting a Rust API. A binary does not have a Rust API. So rustdoc is not really meant for documenting binaries.

Where the friction really comes is when you want to produce *internal* documentation. This is a concern for libraries too, and that's why `--document-private-items` exists. But the way that rustdoc goes about generating this information ends up being tough for binaries, because it's oriented around lbiraries, by nature.

**Comment by kornelski** [comment]
I think the concern of running (doc)tests could be separate from concern of documenting things. Some tests like `compile_fail` just have to live in a doccomment, regardless whether one wants to generate documentation for a binary or not.

To look at it from a different perspective: tests are run by `cargo test --all`, not `cargo doc`. The fact that rustdoc is involved in there is an implementation detail. If you ignore how the tests are extracted, in the grand scheme it's just `cargo test --all` failing to run all tests.

**Comment by dogweather** [comment]
@joshtriplett Take a look at how Elixir does doc tests: in the form of an interactive REPL session:

```elixir
defmodule MyModule do
  @doc """
  Given a number, returns true if the number is even and else otherwise.

  ## Example
    iex> MyModule.even?(2)
    true
    iex> MyModule.even?(3)
    false
  """
  def even?(number) do
    rem(number, 2) == 0
  end
end
```

This would be a big change, and implies a REPL that everyone is familiar with. But it's a very low-friction way to write tests. Maybe as a custom testing framework when that hits stable...

**Comment by samwightt** [comment]
Just hit this roadblock today with a smaller bin project that I was working on and it was very annoying. Doc tests should be standard across all project types, not just library.

**Comment by huntc** [comment]
I'm seeing this issue cropping up when I *do* have a library, albeit a static library...:

```
% cargo test --doc --verbose      
warning: doc tests are not supported for crate type(s) `staticlib` in package `mylib`
error: no library targets found in package `mylib`
```

Should I raise this as a separate issue? Thanks.

**Comment by rvolgers** [comment]
Just ran into this today.

To the people explaining why this is tricky to implement in a consistent way: I hear you, and I understand.

To the people who (seem to?) be arguing that this is good and normal and expected: you are seriously wrong.

Things that normally work and then in another context quietly *don't do anything* are really bad papercuts.

While I understand there needs to be some guiding principle behind rustdoc, the idea that it is for documenting Rust APIs and therefore it doesn't make sense for binaries is not a helpful one. By that logic we could ban doc comments entirely in bin crates, at least that would be a consistent position.

TLDR: doc tests not working may be a tricky bug, but it sure isn't a feature

**Comment by GuillaumeGomez** [comment]
> TLDR: doc tests not working may be a tricky bug, but it sure isn't a feature

No one said it was a feature, we just said it was out of rustdoc scope. Having doc tests in a binary doesn't sound crazy to me, it's simply something that we need to agree upon and then to update rustdoc source code so that it can work there too (which might be quite tricky...).

**Comment by ganzhi** [comment]
Just hit it today. Understand it was out of rustdoc scope till now. How can we make it part of rustdoc scope? :)

**Comment by slanden** [comment]
> It sounds weird to have documentation examples in a bin target, don't you think? There is no documentation to generate so why would there be documentation's code to run?

Just my two cents, I think having tests to validate documented examples in a bin target is just as important. After all, contributors of the code need to know how it works. New contributors or old returning after months away.

Somebody might've changed the code but forgot to update the docs. Having dedicated tests of course is better and helps, but they're different; I like having both.

**Comment by jyn514** [comment]
This is not actually a rustdoc bug: you can run `rustdoc --test` on a binary just fine:
```
$ cat src/main.rs
/// ```
/// assert!(true);
/// ```
fn main() {}
$ rustdoc --test src/main.rs 

running 1 test
test src/main.rs - main (line 1) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.12s
```
The problem is Cargo doesn't run `rustdoc --test` when you run `cargo test --bin` or `cargo test --all-targets`.

@ehuss could you move this to rust-lang/cargo?

**Comment by ehuss** [comment]
I don't think this is a cargo issue, as I think there are some fundamental issues with how doctests link with the surrounding code, as described above at https://github.com/rust-lang/rust/issues/50784#issuecomment-605521414.  It would be feasible to run doctests on binaries as you mentioned, but not very useful since you wouldn't be able to access any of the code within the binary.

My opinion is that it would be good for someone to write a proposal for a different way for doctests to work.  I think embedding them as regular unit tests is a compelling option, as it would solve the issue of working with binaries, working with private items (#60820), and improving performance (#75341).  It won't be easy and it will likely need to be opt-in (with possibly a per-test opt-out), but I think it is worthwhile.

**Comment by EvanCarroll** [comment]
This is a real gotcha. It's not clear to me why doc tests should not run on a binary, when the same code (in `src/main.rs`) generates code docs with `cargo doc`

**Comment by kupiakos** [comment]
Got bit by this today: I have a private `util` module deep in a crate. However, I'd still like to provide testable examples to the in-crate users of this module. There's no way to do this, since rustdoc was initially designed with the idea of "document this public Rust API" and not "document this Rust code". `rustdoc` needs a paradigm shift here to make doctests consistently useful. Documenting monolithic internals is useful too :)

I begin every one of my doctests with `# use crate_name::path::to::item::ItemUnderTest`. I understand why this is currently required, but requiring it of every doc test *is a papercut*, and is the key reason why it's impossible to doctest in a private module. Perhaps we can resolve paths inside doctests to be relative to the item under test? Or implicitly expose the item under test? Or somehow grant an exception for access to private items when accessing the item under test in a doctest?

**Comment by Kromey** [comment]
This bit me today, I'm working on a large (well, will be) project, and for both myself and potential future contributors I've been using rustdoc with `--document-private-items` to create internal documentation. I threw a simple doctest in one, but because it's inside a private module it of course failed to compile.

Not knowing anything about how they work, would it be possible to have rustdoc generate *unit tests* from doctests, rather than separate bin crates? I realize this would probably be a breaking change to existing (and functional) doctests, not to mention may violate the original intent of documenting public APIs, so perhaps in the same way we can add attributes such as `ignore` or `compile_fail`, could a `unit_test` attribute be added to doctests we want to use this behavior?

My current workaround is to add the `ignore` tag to my code blocks, so that at least I can still provide example code in the documentation, but of course this largely defeats the purpose of having doctests in the first place, which is to help ensure that code and documentation remain in sync.

**Comment by bryanlarsen** [comment]
As a workaround it's fairly straightforward to add a lib.rs and a [lib] entry to your Cargo.toml to make your crate both a binary and a library.

**Comment by GuillaumeGomez** [comment]
Already suggested here: https://github.com/rust-lang/rust/issues/50784#issuecomment-653895749. :wink:

**Comment by AntoniosBarotsis** [comment]
I also stumbled upon this. 

If you happen to have both a `main` and a `lib`, be sure to specify them in your `cargo.toml` like so:

```toml
[[bin]]
name = "main"
path = "src/main.rs"

[lib]
name = "lib"
path = "src/lib.rs"
```

Got this from [here](https://stackoverflow.com/a/57767413/12756474).

**Comment by jyn514** [comment]
https://github.com/dtolnay/trybuild is a library for testing compile errors. I think it's unlikely it will become part of libtest itself.

**Comment by DragonDev1906** [comment]
> Not knowing anything about how they work, would it be possible to have rustdoc generate unit tests from doctests, rather than separate bin crates? I realize this would probably be a breaking change to existing (and functional) doctests, not to mention may violate the original intent of documenting public APIs, so perhaps in the same way we can add attributes such as ignore or compile_fail, could a unit_test attribute be added to doctests we want to use this behavior?

I'd like to point out this comment by @Kromey:
- We can not convert **all** doctests to unit tests, because that would mean `compile_fail` would not work.
- But it should be possible to add the attributes `unit_test` and `unit_test_should_fail` (or something like that), which converts only this single example to a unit test.

While it would be impossible to use the `compile_fail` attribute for private functions (use `ignore` as a fallback or keep `compile_fail` even if it always fails compilation due to using private methods), this would at least allow us to use doctests for private functions or in private modules that correctly compile (i.e. everything usually done without attributes or `should_panic`. At the same time this would not be a breaking change, since these examples currently do not compile anyways.

For those tests we could automatically `use super::*` as with unit tests and it would make sense to do so because they should behave like a unit test.

Alternatively, this could be done automatically (without a `unit_test` attribute) for all doctests in private modules, where it is already impossible to run the examples if they use private functions. This might be a breaking change, since existing doctests for private modules/functions that do **not** use these functions would be run in a different context.

If you ask me, having the ability to execute doctests in private modules iff they are supposed to compile is a big benefit for internal documentation of private modules in a library, even if it has the limitation that the example must compile. Especially when this can be done without a breaking change or in a way where adding the attribute can be automated later. It is really annoying that we have to add `ignore`  to all examples in private modules just because we can not import the function the example is about for the doctest.

**Comment by TheButlah** [comment]
This bit me today. I am writing a backend service, which has a JSON api. I wanted to document an example of valid JSON, for someone reading the code.

Conceptually this *should* only need to be a `bin` crate - its an application. But I still have a need to document its API. It has an API because its exposing HTTP endpoints, and that is a form of API.

It would be good if I didn't have to resort to making a trivial `lib` crate with a `main()` function (which is what I used to do for all applications). That is not a good developer experience.

**Comment by ChristopherPerry6060** [comment]
This caught me today. It seems like very unintuitive behavior. Since I am still very new to the language, the docs are mostly for my own benefit and sanity. I know there is an unstable feature to scrape examples, but I really wish I could just write my examples outside of markdown blocks while still having them be within the same file of what they are documenting.

**Comment by Ofenhed** [comment]
I will humbly admit I'm not well versed in rustc or cargo internals, so I may be way off. I would still like to make a few comments based on what I would want and which behaviors I would expect as a developer when running `cargo test`:
1. Public facing documentation targets in libraries should really work as they do today. Giving them access to `super::` would be detrimental to the quality of documentation.
2. Private facing documentation targets (such as private functions in libraries or any function in a binary) could be converted to anonymous submodules where they are documented. This would give us access to `super::`.
    1. For documentation tests marked as `compile_fail`, the `use` statement should be conditional. Maybe behind a feature flag, maybe conditionally generated.
    2. Fully private documentation targets *could* be converted to tests in the same module, as the declared `use` statements are valid for all functions which can call this function. This is obviously not true for tests marked with `compile_fail`.
    3. Tests for documentation targets which are not fully private should never just be converted to tests in the same module. I think that declaring usage makes the documentation clearer.
    4. The anonymous submodules could be confusing in compiler error text (file paths), so we would need to process that output. Can we do this?
3. I guess that for this to work we would also need to inject code into `main.rs`/`lib.rs` to publish the anonymous test modules to the function which runs the tests?

**Comment by daniel-pfeiffer** [comment]
> It sounds weird to have documentation examples in a bin target, don't you think?

A rather nerdy answer, don't you think? I.e. looking at the world through your binoculars, instead of taking the client's perspective. And Rust-programmers in a sence are clients of the language.

When we start, we get shown all the shiny sides, like inline documentation that doubles as unit-tests – wow! Been putting off till now, why it's not working. To my great dismay, I find out that it's only for the minority of lib-providers, not for the majority who write bins to solve real world problems. (No belittling intended – libs are valuable!)

I get the technical reasons, and I'm sympathetic that someone needs to put in the effort. Kudos to those who do! Yet for now, we're stuck with a broken paradigm: lack of consistency in how to document and unit-test functions (and macros!) And yes, also in a bin, every single one should be documented (and as far as possible tested.)

And it hampers my desire to advocate for Rust. Just too many rough edges, due to random decisions like having rustdoc take care of the testing. Instead I would put this into a compiler test-plugin, which collects all these snippets and runs them at the end of their scope. Probably far more efficient, too…

**Comment by khionu** [comment]
@daniel-pfeiffer People aren't always going to have the variety of perspectives at the early stages of a given exploration. We should be empathetic to that people come from all walks of life and may need some help to understand other perspectives. Your response here was very much not how we should help each other understand things, and that's putting aside that you're replying to something from almost 5 years ago. In addition to the empathy, please watch for context.

Lastly, a lot of your response was loaded with emotional context that didn't do anything for the conversation. All feelings are valid, but how you act on them matters. This wasn't the place to vent about your feelings with how contributions are done.

**Comment by JeffThomas** [comment]
Yeah this caught me today too. I'm writing something that's not intended to be used as-is (such as a library) but is meant to be forked and expanded on for the users special case. Working examples would be very valuable, I think. I understand that this is a tricky issue but since people are still commenting on it some 4 years after it was first reported, maybe it can be moved up on in priority? 

Having functioning examples in the documentation is such a powerful and useful feature it's a shame that it's limited in such a way.

**Comment by GuillaumeGomez** [comment]
As suggested in the first comments, it could already work by providing a small fix in cargo, which would be the simplest way to achieve this feature. You can try your luck there directly.

**Comment by modulok** [comment]
I was bit by this today too. I'm in favor of doctests just running, binary or not. It can be very useful when teaching Rust to be able to put a doctest in a binary. Given the number of people just here who have hit this - and bothered to comment - and the number of years this has bitten folks, it'd probably more of an expected behavior if they just always run, even in binaries.

**Comment by vi** [comment]
If it is tricky to _run_ those doctests, maybe it can be easier to detect them and warn.

```rust
/// ```
///   2+2
/// ```
fn main() {
    println!("Hello, world!");
}
```

```
warning: doctests do not work in [[bin]] targets or in private items
 --> src/main.rs:1:4
```

**Comment by modulok** [comment]
Agreed. That would be helpful until an always-run-tests feature can (if possible and desired)  be properly designed and implemented. I doubt implementing such a warning could break anyone's existing code.

**Comment by velllu** [comment]
I am writing my backend in rust, and I would love to write doctests for the sql-related stuff, but my project is a binary :thinking:

I really cannot see a reason to not want this, it does not change any existing programs.

**Comment by WyvernIXTL** [comment]
The best part is, that in [rust by example](https://doc.rust-lang.org/rust-by-example/testing/doc_testing.html) there is no mention of doctest not working with bin targets:
*"Rust takes care about correctness, so these code blocks are compiled and used as documentation tests."*
*"Code blocks in documentation are automatically tested when running the regular cargo test command"*

It should just work.
*(A lengthy bump.)*

**Comment by Ziothh** [comment]
This also bit me a while ago and every couple of months I'm searching the issues for progress because I do think this could greatly help people understand other peoples code and intentions. 

These are my main pain points:
- Having both a `main.rs` and a `lib.rs` is a minor pain point but you shouldn't have to. It could make external people think you're sharing lib code or publishing it to crates.io.
- I'd guess that most rust projects are binaries so the majority of projects don't directly benefit from doctests.
- Even if you have a lib, private function doctests don't work (I know this isn't trivial to implement atm). 
  - Because I'm not a library author my private and public functions benefit the same amount from doctests.
  - Rust shouldn't force you tot make functions public just to be testable in the way that you prefer.
- Having great examples for internal code helps team members to better understand each others intentions.
- I often find it easier to understand open source libs instead of bins because they're incentivized to write examples.

**Comment by nm-remarkable** [comment]
2024 reporting in! Was very confused why my tests were not running after following [rust by example](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html). The part where it says `For example, you could write this into your lib.rs to test your README as part of your doctests:` makes it sounds optinal, and there is no documentation that documentation tests do not work at all with `bin` (why I found this issue in the first place).

On a cheerful note, it's a good time to drop [this article](https://news.ycombinator.com/item?id=38797908).

_(Once more a lenghty bump on this topic)_

**Comment by williamareynolds** [comment]
I spent an hour trying to figure out why my doctests wouldn't run. So happy I finally found this issue.

**Comment by GuillaumeGomez** [comment]
Little update for everyone interested: discussions on how to fix this issue (re)started recently. As you might have guessed, it's a tricky issue to solve as it requires to completely change how doctests work. The "fun" part is allowing them to work on internal APIs (the current issue) and also allow them to work as if your crate is a dependency to showcase how to use an API (the original intent and current way they work).

**Comment by GuillaumeGomez** [comment]
Just to keep everyone up-to-date: since https://github.com/rust-lang/rust/pull/126245 was merged, the first step is now done. Here is the plan I currently have in mind for allowing this feature to work: just like currently, `rustdoc` will process all doctests. However, all private items (not reexported or not accessible outside of the current crate, so all items in a binary crate) will now be transformed using the same logic as unit tests (`#[test]`).

How does it work? If any doctest matches the parameter listed above, `rustdoc` will call `rustc` with a flag telling `rustc` to convert these doctests into something similar to unit tests and to run them. It means they will be compiled alongside their crate, giving access to the module where they are declared.

Now what happens if you want a doctest on a public/reexported item to still have access to local code? For that I suggest adding a new code block attribute named `unit_test` (name is very much up to debate!). These doctests will be treated the same as the ones on private items.

So from this point, the only issue I can foresee in the implementation is for doctests on impl items. It will mean that the test will need to be generated outside of the impl block. I'll experiment a bit around that to see how to make it possible.

I think that's it. Work on it began. A bit more patience folks. Binacry crate/private items doctests are coming!

**Comment by 9-FS** [comment]
Is there any update on this in 2025?

**Comment by GuillaumeGomez** [comment]
We now have a plan:

* Run all compiler passes in rustdoc
* Convert doctests as code with `#[test]` that then is passed to the compiler to generate the binary.

Hopefully I didn't mess up the explanations. It's quite a lot of work and unlikely we'll get there before the end of the year.

**Comment by 9-FS** [comment]
Bro answered faster than I could get my coffee, I love you.
